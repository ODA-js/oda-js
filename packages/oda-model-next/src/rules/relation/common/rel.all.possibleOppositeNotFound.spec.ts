import Rule from './possibleOppositeNotFound';

describe('rule', () => {
  let rule: Rule;
  beforeAll(() => {
    rule = new Rule();
  });

  it('fixable', () => {
    const updateWith = jest.fn();
    const result = rule.validate({
      relation: {
        opposite: false,
        ref: {
          entity: true,
        },
        updateWith,
      },
      entity: {
        name: true,
      },
      field: {
        name: true,
      },
      package: {
        items: {
          get: jest.fn().mockReturnValueOnce({
            modelType: 'entity',
            fields: {
              values: jest.fn().mockReturnValueOnce([
                {
                  name: true,
                  relation: {
                    ref: {
                      entity: true,
                      field: true,
                    },
                  },
                },
              ]),
            },
          }),
        },
      },
    } as any);
    expect(updateWith).toBeCalledWith({ opposite: true });
    expect(result).toMatchSnapshot();
  });

  it('throw regular more than one opposite', () => {
    const updateWith = jest.fn();
    const result = rule.validate({
      relation: {
        opposite: false,
        ref: {
          entity: true,
        },
      },
      entity: {
        name: true,
      },
      field: {
        name: true,
      },
      package: {
        items: {
          get: jest.fn().mockReturnValueOnce({
            modelType: 'entity',
            fields: {
              values: jest.fn().mockReturnValueOnce([
                {
                  relation: {
                    ref: {
                      entity: true,
                      field: true,
                    },
                  },
                },
                {
                  relation: {
                    ref: {
                      entity: true,
                      field: true,
                    },
                  },
                },
              ]),
            },
          }),
        },
      },
    } as any);
    expect(updateWith).not.toBeCalled();
    expect(result).toMatchSnapshot();
  });

  it('critics no possible opposite', () => {
    const updateWith = jest.fn();
    const result = rule.validate({
      relation: {
        opposite: false,
        ref: {
          entity: true,
        },
      },
      entity: {
        name: true,
      },
      field: {
        name: true,
      },
      package: {
        items: {
          get: jest.fn().mockReturnValueOnce({
            modelType: 'entity',
            fields: {
              values: jest.fn().mockReturnValueOnce([]),
            },
          }),
        },
      },
    } as any);
    expect(updateWith).not.toBeCalled();
    expect(result).toMatchSnapshot();
  });

  it('throw belongsToMany one fixable opposite', () => {
    const updateWith = jest.fn();
    const result = rule.validate({
      relation: {
        modelType: 'relation',
        verb: 'BelongsToMany',
        opposite: false,
        ref: {
          entity: true,
        },
        updateWith,
      },
      entity: {
        name: true,
      },
      field: {
        name: true,
      },
      package: {
        items: {
          get: jest.fn().mockReturnValueOnce({
            modelType: 'entity',
            fields: {
              values: jest.fn().mockReturnValueOnce([
                {
                  name: true,
                  relation: {
                    modelType: 'relation',
                    verb: 'BelongsToMany',
                    ref: {
                      entity: true,
                      field: true,
                    },
                  },
                },
              ]),
            },
          }),
        },
      },
    } as any);
    expect(updateWith).toBeCalledWith({ opposite: true });
    expect(result).toMatchSnapshot();
  });

  it('not throw 1', () => {
    const updateWith = jest.fn();
    const result = rule.validate({
      relation: {
        opposite: true,
        ref: {
          entity: true,
        },
        updateWith,
      },
    } as any);
    expect(updateWith).not.toBeCalled();
    expect(result).toMatchSnapshot();
  });

});
