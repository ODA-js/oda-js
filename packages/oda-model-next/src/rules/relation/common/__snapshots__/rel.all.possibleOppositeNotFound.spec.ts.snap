// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule critics no possible opposite 1`] = `
Array [
  Object {
    "message": "no possible opposite",
    "result": "critics",
  },
]
`;

exports[`rule fixable 1`] = `
Array [
  Object {
    "message": "found one possible opposite. assigned.",
    "result": "fixable",
  },
]
`;

exports[`rule not throw 1 1`] = `Array []`;

exports[`rule throw belongsToMany one fixable opposite 1`] = `
Array [
  Object {
    "message": "found one possible opposite. assigned.",
    "result": "fixable",
  },
]
`;

exports[`rule throw regular more than one opposite 1`] = `Array []`;
